lib/src/ssh_signing.rs

```diff
@@ -0,0 +1,305 @@
+            return Ok(None);
+        }
+        Ok(Some(principal))
+    }
+}
+
+impl SigningBackend for SshBackend {
+    fn name(&self) -> &str {
+        "ssh"
+    }
+
+    fn can_read(&self, signature: &[u8]) -> bool {
+        signature.starts_with(b"-----BEGIN SSH SIGNATURE-----") && self.allowed_signers.is_some()
+    }
+
+    fn sign(&self, data: &[u8], key: Option<&str>) -> Result<Vec<u8>, SignError> {
+        let Some(key) = key else {
+            return Err(SshError::MissingKey.into());
+        };
+
+        let (pub_key_path, _file) = ensure_key_as_file(key)?;
+        let mut command = self.create_command();
+
+        command
+            .arg("-Y")
+            .arg("sign")
+            .arg("-f")
+            .arg(pub_key_path.as_os_str())
+            .arg("-n")
+            .arg("git");
+
+        Ok(run_command(&mut command, data)?)
+    }
+
+    fn verify(&self, data: &[u8], signature: &[u8]) -> Result<Verification, SignError> {
+        let mut signature_file = tempfile::Builder::new()
+            .prefix(".jj-ssh-sig-")
+            .tempfile()
+            .map_err(SshError::Io)?;
+        signature_file.write_all(signature).map_err(SshError::Io)?;
+        signature_file.flush().map_err(SshError::Io)?;
+
+        let signature_file_path = signature_file.into_temp_path();
+
+        let principal = self.find_principal(&signature_file_path)?;
+
+        let mut command = self.create_command();
+
+        match principal {
+            None => {
+                command
+                    .arg("-Y")
+                    .arg("check-novalidate")
+                    .arg("-s")
+                    .arg(signature_file_path)
+                    .arg("-n")
+                    .arg("git");
+
+                let result = run_command(&mut command, data);
+
+                match result {
+                    Ok(_) => Ok(Verification::new(
+                        SigStatus::Unknown,
+                        None,
+                        Some("Signature OK. Unknown principal".into()),
+                    )),
+                    Err(_) => Ok(Verification::new(SigStatus::Bad, None, None)),
+                }
+            }
+            Some(principal) => {
+                let allowed_signers = self.get_allowed_signers()?;
+
+                command
+                    .arg("-Y")
+                    .arg("verify")
+                    .arg("-s")
+                    .arg(signature_file_path)
+                    .arg("-I")
+                    .arg(&principal)
+                    .arg("-f")
+                    .arg(&allowed_signers)
+                    .arg("-n")
+                    .arg("git");
+
+                let result = run_command(&mut command, data);
+
+                match result {
+                    Ok(_) => Ok(Verification::new(SigStatus::Good, None, Some(principal))),
+                    Err(_) => Ok(Verification::new(SigStatus::Bad, None, Some(principal))),
+                }
```

Comment:
nit: you could extract repetition like this
```suggestion
                let status = match result {
                    Ok(_) => SigStatus::Good,
                    Err(_) => SigStatus::Bad,
                };
                Ok(Verification::new(status, None, Some(principal)))
```